internal/controller/postgrescluster/volumes.go

```diff
@@ -122,6 +131,557 @@ func (r *Reconciler) observePersistentVolumeClaims(
+					Image:        config.PostgresContainerImage(cluster),
+					Name:         naming.JobMovePGDataDir,
+					VolumeMounts: []corev1.VolumeMount{postgres.DataVolumeMount()},
+				}},
+				SecurityContext: postgres.PodSecurityContext(cluster),
+				// Set RestartPolicy to "Never" since we want a new Pod to be
+				// created by the Job controller when there is a failure
+				// (instead of the container simply restarting).
+				RestartPolicy:                v1.RestartPolicyNever,
+				AutomountServiceAccountToken: initialize.Bool(false),
+				Volumes: []v1.Volume{{
+					Name: "postgres-data",
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							ClaimName: cluster.Spec.DataSource.ExistingVolumes.
+								ExistingPGDataVolume.PVCName,
+						},
+					}},
+				},
+			},
+		},
+	}
+	moveDirJob.Spec = *jobSpec
+
+	// set gvk and ownership refs
+	moveDirJob.SetGroupVersionKind(batchv1.SchemeGroupVersion.WithKind("Job"))
+	if err := controllerutil.SetControllerReference(cluster, moveDirJob,
+		r.Client.Scheme()); err != nil {
+		return true, errors.WithStack(err)
+	}
+
+	// server-side apply the backup Job intent
+	if err := r.apply(ctx, moveDirJob); err != nil {
+		return true, errors.WithStack(err)
+	}
+
+	return true, nil
+}
+
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=create;patch;delete
+
+// reconcileMoveWalDir creates a Job to move the provided pg_wal directory
+// in the given volume to the expected location before the PostgresCluster is
+// bootstrapped. It returns any errors and a boolean indicating whether the
+// main control loop should continue or return early to allow time for the job
+// to complete.
+func (r *Reconciler) reconcileMoveWALDir(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, moveJobs *batchv1.JobList) (bool, error) {
+
+	moveDirJob := &batchv1.Job{}
+	moveDirJob.ObjectMeta = naming.MovePGWALDirJob(cluster)
+
+	// check for an existing Job
+	for i := range moveJobs.Items {
+		if moveJobs.Items[i].Name == moveDirJob.Name {
+			if jobCompleted(&moveJobs.Items[i]) {
+				// if the Job is completed, return as this only needs to run once
+				return false, nil
+			}
+			if !jobFailed(&moveJobs.Items[i]) {
+				// if the Job otherwise exists and has not failed, return and
+				// give the Job time to finish
+				return true, nil
+			}
+		}
+	}
+
+	moveDirJob.ObjectMeta.Annotations = naming.Merge(cluster.Spec.Metadata.
+		GetAnnotationsOrNil())
+	labels := naming.Merge(cluster.Spec.Metadata.GetLabelsOrNil(),
+		naming.DirectoryMoveJobLabels(cluster.Name),
+		map[string]string{
+			naming.LabelMovePGWalDir: "",
+		})
+	moveDirJob.ObjectMeta.Labels = labels
+
+	script := fmt.Sprintf(`echo "Preparing cluster %s volumes for PGO v5.x"
+    echo "pg_wal_pvc=%s"
+    echo "Current PG WAL directory volume contents:"
+    ls -lh "/pgwal"
+    echo "Now updating PG WAL directory..."
+    [ -d "/pgwal/%s" ] && mv "/pgwal/%s" "/pgwal/%s-wal"
```

Comment:
I dug into this a bit to double check. The `/pgwal/clustername-wal` directory is expected by the [startup script](https://github.com/CrunchyData/postgres-operator/blob/master/internal/postgres/config.go#L158-L167). Once this script completes, the final location will be the expected `/pgwal/pg13_wal` directory. That is why the existing v4 volumes don't require the move Job to be run.

Reply:
> I dug into this a bit to double check. The `/pgwal/clustername-wal` directory is expected by the [startup script](https://github.com/CrunchyData/postgres-operator/blob/master/internal/postgres/config.go#L158-L167).

The startup script expects a number and underscore:

https://github.com/CrunchyData/postgres-operator/blob/eb994a36ba22c61535174c50cb6fac2e2e3b34d5/internal/postgres/config.go#L107-L112 https://github.com/CrunchyData/postgres-operator/blob/eb994a36ba22c61535174c50cb6fac2e2e3b34d5/internal/postgres/config.go#L166

<br/>

> Once this script completes, the final location will be the expected `/pgwal/pg13_wal` directory. That is why the existing v4 volumes don't require the move Job to be run.

🤔 I believe we're hitting [this scenario](https://github.com/CrunchyData/postgres-operator/blob/eb994a36ba22c61535174c50cb6fac2e2e3b34d5/internal/postgres/config_test.go#L261-L262) handled by the startup script:

```
lrwxrwxrwx. /pgdata/pg13/pg_wal -> /pgwal/clustername-wal
drwx------. /pgwal/clustername-wal
```

It works because the WAL volume is mounted at the same location in v4 and v5. The `pg_wal` symlink on the data volume points to an absolute path that happens to exist.

If the volumes were previously mounted somewhere else, then the symlink would point nowhere. For example,

```
# in the past
lrwxrwxrwx. /mnt/data-volume/datadir/pg_wal -> /mnt/wal-volume/waldir
drwx------. /mnt/wal-volume/waldir

# during our startup
lrwxrwxrwx. /pgdata/pg13/pg_wal -> /mnt/wal-volume/waldir
drwx------. /pgwal/waldir
```

🤔 We probably need to mount the data volume and WAL volume into a single/same job so the name of the data directory and the contents of the symlink inside it can be corrected together.

The spec for my example above would be:
```yaml
      existingPGDataVolume: { directory: datadir }
      existingPGWALVolume:  { directory: waldir }
```

And the script would need to:

1. Rename `/pgdata/datadir` to `/pgdata/pg13`
2. Rename `/pgwal/waldir` to `/pgwal/pg13_wal`
3. Link from `/pgdata/pg13/pg_wal` to `/pgwal/pg13_wal`.